Initial docs for file access auth feature (#1017)
google/santa: docs/deployment/file-access-auth.md

---
title: File Access Authorization
parent: Deployment
nav_order: 4
---

# (BETA) File Access Authorization

> > **IMPORTANT:** This feature is currently in beta. Configuration and log formats are subject to change.

Reviewer: ```suggestion
> **IMPORTANT:** This feature is in beta. Configuration and log formats are subject to change.
```

Reviewer: Perhaps remove "currently" it doesn't really add anything. The beta label will be removed once the feature development is at a particular point anyway. 

In the note, "in beta" is fine, but another alternative is:

> **IMPORTANT:** This feature is in development. Configuration and log formats are subject to change.

"development" feels more active and provides more context when used with the "(BETA)" label. But, totally up to you, either way is good.

> > **IMPORTANT:** This feature is currently in beta. Configuration and log formats are subject to change.

Reviewer: If there are a couple of things you'd like to call out especially for issues that are to be fixed, you could group under a ## Known Limitations header. This might be a bit excessive here, but worth considering. Providing even more transparency with a linked GH issue means that users can also follow the issue if they wish. Just a thought!

> > **IMPORTANT:** This feature is currently in beta. Configuration and log formats are subject to change.

Author: I like the idea of a "Known Limitations" section, but I wasn't able to make it look right to me. Things are currently ordered in a way that introduces concepts and groups the limitations next to each other. Breaking that out would remove that grouping and potentially make it more difficult to discover pertinent information.

Re: calling out the delay - I expanded a bit up top about the update interval key existing and added and link to the section with more details.

> **IMPORTANT:** This feature is only supported on macOS 13 and above.

File Access Authorization allows admins to configure Santa to monitor filesystem paths for potentially unwanted access and optionally deny the operation.

## Enabling the Feature

> To enable this feature, the `FileAccessPolicyPlist` key in the main Santa configuration must contain the path to a configuration file (format specified below).

Reviewer: When you say "format specified below" here, do you mean the ## Configuration section or the ## Example Configuration? Might be worth linking to either, e.g.:

"... a configuration file. See the format specified in the [Example Configuration](#example-configuration)."

## Configuration

| Key                 | Parent       | Type       | Required | Default    | Description |
| ------------------- | ------------ | ---------- | -------- | ---------- | ----------- |
| `Version`           | `<Root>`     | String     | Yes      | N/A        | Version of the configuration. Will be reported in events. |
| `WatchItems`        | `<Root>`     | Dictionary | No       | N/A        | The set of configuration items that will be monitored by Santa. |

> | `<Name>`            | `WatchItems` | Dictionary | No       | N/A        | A unique name that identifies a single watch item rule. This value will be reported in events. The name must be a legal C identifier (e.g. must conform to the regex `[A-Za-z_][A-Za-z0-9_]*`). |

Reviewer: ```suggestion
| `<Name>`            | `WatchItems` | Dictionary | No       | N/A        | A unique name that identifies a single watch item rule. This value will be reported in events. The name must be a legal C identifier (e.g., must conform to the regex `[A-Za-z_][A-Za-z0-9_]*`). |
```

Reviewer: There's a mix of e.g. or i.e. with following commas. I'm adding the comma for consistency as that's what I see in other docs (feel free to go either way though).

| `Paths`             | `<Name>`     | Array      | Yes      | N/A        | A list of either String or Dictionary types that contain path globs to monitor. String type entires will have default values applied for the attributes that can be manually set with the Dictionary type. |
| `Path`              | `Paths`      | String     | Yes      | None       | The path glob to monitor. |
| `IsPrefix`          | `Paths`      | Boolean    | No       | False      | Whether or not the path glob represents a prefix path. |
| `Options`           | `<Name>`     | Dictionary | No       | N/A        | Customizes the actions for a given rule. |

> | `AllowReadAccess`   | `Options`    | Boolean    | No       | False      | If true, indicates the rule will NOT be applied to actions that are read-only access (e.g. opening a watched path for reading, or cloning a watched path). If false, the rule will apply both to read-only access and access that could modify the watched path. |

Reviewer: ```suggestion
| `AllowReadAccess`   | `Options`    | Boolean    | No       | False      | If true, indicates the rule will **not** be applied to actions that are read-only access (e.g. opening a watched path for reading, or cloning a watched path). If false, the rule will apply both to read-only access and access that could modify the watched path. |
```

> | `AllowReadAccess`   | `Options`    | Boolean    | No       | False      | If true, indicates the rule will NOT be applied to actions that are read-only access (e.g. opening a watched path for reading, or cloning a watched path). If false, the rule will apply both to read-only access and access that could modify the watched path. |

Reviewer: ```suggestion
| `AllowReadAccess`   | `Options`    | Boolean    | No       | False      | If true, indicates the rule will NOT be applied to actions that are read-only access (e.g., opening a watched path for reading, or cloning a watched path). If false, the rule will apply both to read-only access and access that could modify the watched path. |
```

| `AuditOnly`         | `Options`    | Boolean    | No       | True       | If true, operations violating the rule will only be logged. If false, operations violating the rule will be denied and logged. |
| `Processes`         | `<Name>`     | Array      | No       | N/A        | A list of dictionaries defining processes that are allowed to access paths matching the globs defined with the `Paths` key. For a process performing the operation to be considered a match, it must match all defined attributes of at least one entry in the list. |

> | `BinaryPath`        | `Processes`  | String     | No       | None       | A path literal an instigating process must be executed from. |

Reviewer: ```suggestion
| `BinaryPath`        | `Processes`  | String     | No       | None       | A path literal that an instigating process must be executed from. |
```

For example, consider the configured rules and paths:
```
RULE_1: /tmp/foo     [IsPrefix=true]
RULE_2: /tmp/foo.txt [IsPrefix=false]
RULE_3: /tmp         [IsPrefix=true]
```
The following table demonstrates which rule will be applied for operations on a given path:
| Operation Path   | Rule Applied | Reason |
| ---------------- | ------------ | ------ |
| /tmp/foo         | `RULE_1`     | Matches prefix, more specific than `RULE_3` |
| /tmp/foo/bar     | `RULE_1`     | Matches prefix, more specific than `RULE_3` |
| /tmp/bar         | `RULE_3`     | Matches prefix |
| /tmp/foo.txt     | `RULE_2`     | Matches literal, more specific than `RULE_1` |
| /tmp/foo.txt.tmp | `RULE_1`     | Matches prefix, more specific than `RULE_3`, literal match doesn't apply |
| /foo             | N/A          | No rules match operations on this path |

If a configuration contains multiple rules with duplicate configured paths, only one rule will be applied to the path. It is undefined which configured rule will be used. Administrators should take care not to define configurations that may have duplicate paths.

### Path Globs

> Configured path globs represent a point in time. That is - path globs are expanded when a configuration is applied to generate the set of monitored paths. Currently this is not a "live" representation of the filesystem. For instance, if a new file or directory is added that would match a glob after the configuration is applied, it is not immediately monitored.

Reviewer: ```suggestion
Configured path globs represent a point in time. That is, path globs are expanded when a configuration is applied to generate the set of monitored paths. Currently, this is not a "live" representation of the filesystem. For instance, if a new file or directory is added that would match a glob after the configuration is applied, it is not immediately monitored.
```

Reviewer: Optional: Remove "Currently" --> "This not a 'live' representation..."

I would avoid using "Currently" since it implies (to me) that at some point a "live" representation of the filesystem would be supported. While that might be true, it's possible to frustrate users with expectations that aren't meant according to what that timeframe could look like to them.

Within the main Santa configuration, the `FileAccessPolicyUpdateIntervalSec` key controls how often any changes to the configuration are applied as well as re-evaluating configured path globs to match the current state of the filesystem.

### Prefix and Glob Path Evaluation

> Combining path globs and the `IsPrefix` key allows configurations greater control over the paths rules should match. Consider the configured path globs:

Reviewer: ```suggestion
Combining path globs and the `IsPrefix` key in a configuration gives greater control over the paths that rules should match. Consider the configured path globs:
```

```
PG_1: /tmp/*         [IsPrefix = false]
PG_2: /tmp/*         [IsPrefix = true]
PG_3: /tmp/          [IsPrefix = true]
PG_4: /tmp/file1.txt [IsPrefix = false]
```

And a filesystem that contains:
```
/
	tmp/
		file1.txt
		file2.txt
		dir1/
			d1_f1.txt
			d1_f2.txt
```

> Now assume the configuration is applied, and moments later a new file (`/tmp/file3_new.txt`) new directory (`/tmp/dir2_new`) are both created:

Reviewer: ```suggestion
Now, assume the configuration is applied, and moments later a new file (`/tmp/file3_new.txt`) and a new directory (`/tmp/dir2_new`) are both created:
```

Reviewer: Check the meaning here, it looked as if the conjunction and article were missing, but I might be misunderstanding the sense here.

> * `PG_1` will match against the two original files within `/tmp` and the one directory `dir1` itself (but not nested contents)

Reviewer: ```suggestion
* `PG_1` will match against the two original files within `/tmp` and the one directory `dir1` itself (but not nested contents).
```

Reviewer: Bullets are complete sentences, adding period to each.

> * `PG_2` will match against the two original files within `/tmp` and the one directory `dir1` (as well as nested contents)

Reviewer: ```suggestion
* `PG_2` will match against the two original files within `/tmp` and the one directory `dir1` (as well as nested contents).
```

> * `PG_3` will match against all original and newly created files and directories within `/tmp` (as well as nested contents)

Reviewer: ```suggestion
* `PG_3` will match against all original and newly created files and directories within `/tmp` (as well as nested contents).
```

> * `PG_4` will only match `/tmp/file1.txt`

Reviewer: ```suggestion
* `PG_4` will only match `/tmp/file1.txt`.
```

### Case Sensitivity

> All configured paths are case sensitive (i.e. paths specified in both the `Paths` and `BinaryPath` configuration keys). The case must match the case of the path as stored on the filesystem.

Reviewer: ```suggestion
All configured paths are case sensitive (i.e., paths specified in both the `Paths` and `BinaryPath` configuration keys). The case must match the case of the path as stored on the filesystem.
```
